### ISSUE TITLE
CentOS RPM builds broken

### ISSUE THREAD

**Issue by rbberger** [cobbler/cobbler#2165]
Trying to build RPMs on CentOS 7 fails for both the 3.0.1 release and the current master (b6aabeb4bbf7c440f17b6734fb72911cd941f156)

These builds were both ran on a vanilla CentOS 7 system and the CentOS 7 Docker container in the repo.

[v3.0.1_tarball_local_build.txt](https://github.com/cobbler/cobbler/files/3725527/v3.0.1_tarball_local_build.txt)
[v3.0.1_tarball_docker_build.txt](https://github.com/cobbler/cobbler/files/3725529/v3.0.1_tarball_docker_build.txt)
[current_master.txt](https://github.com/cobbler/cobbler/files/3725530/current_master.txt)

The current master gives this error:
```
...
cp: cannot stat 'docs/README.suse': No such file or directory
error: Bad exit status from /var/tmp/rpm-tmp.pkpRbM (%doc)

RPM build errors:
    File not found: /usr/src/cobbler/rpm-build/BUILDROOT/cobbler-3.0.1-1.el7.x86_64/usr/share/cobbler/spool
    File not found: /usr/src/cobbler/rpm-build/BUILDROOT/cobbler-3.0.1-1.el7.x86_64/usr/share/man/man1/cobbler.1.gz
    Bad exit status from /var/tmp/rpm-tmp.pkpRbM (%doc)
```

A quick fix is removing the lines for adding spool, the man pages and the README.suse

```
diff --git a/cobbler.spec b/cobbler.spec
index 591015d..8320209 100644
--- a/cobbler.spec
+++ b/cobbler.spec
@@ -315,15 +315,14 @@ fi
 
 # share
 %{_usr}/share/cobbler
-%exclude %{_usr}/share/cobbler/spool
+#%exclude %{_usr}/share/cobbler/spool
 %exclude %{_usr}/share/cobbler/web
 
 # log
 %{_var}/log/cobbler
 
 # documentation
-%doc AUTHORS COPYING README docs/README.suse
-%{_mandir}/man1/cobbler.1.gz
+%doc AUTHORS COPYING README 
 
 
 #
```

The proper way would probably be to add the Sphinx build of the man pages to `setup.py` and update the spec file to include `cobbler-cli.1`,  `cobbler.conf.5`, and  `cobblerd.1`.

**Comment by rbberger** [comment]
Another problem is the version of `mod_wsgi`. CentOS 8 comes with a Python 3 enabled `mod_wsgi`. The version provided by CentOS 7 is compiled for Python 2. There are several options to get a Python 3.6 mod_wsgi for CentOS 7:

### Option A: Software Collections packages
Enable [Software Collections](https://centos.pkgs.org/7/centos-sclo-rh-x86_64/rh-python36-mod_wsgi-4.5.17-2.el7.x86_64.rpm.html) for CentOS 7 and get the `rh-python36-mod_wsgi` package and adjust the httpd module configuration.

### Option B: Recompile mod_wsgi
Install `httpd-devel` and compile `mod_wsgi` directly using the pip package

```bash
yum remove mod_wsgi
yum install httpd-devel
pip3.6 install mod_wsgi
/usr/local/bin/mod_wsgi-express install-module
echo 'LoadModule wsgi_module "modules/mod_wsgi-py36.cpython-36m-x86_64-linux-gnu.so"' > /etc/httpd/conf.modules.d/10-mod_wsgi-py36.conf
```

To make the RPM build work with this solution, one would have to remove the `mod_wsgi` requirement on the RPM builds and add then pip install in the %pre section of the spec file, similar to the Cheeta3 workaround.

**Comment by SchoolGuy** [comment]
@opoplawski I guess this is you domain. I have to pass because I don't know anything about RPM building on Redhat systems...

**Comment by watologo1** [comment]
@rbberger : Centos 8 is not an option?
cobbler 3.x was mostly about Python3 enabling, cutting off Python 2 requirements.
Not sure how Centos build system works. Maybe it would make sense to enable things for Centos 8 and then try to fulfill missing requirements for Cenos 7.
For SUSE in the OBS (OpenSUSE Buildservice) it would be to open a cobbler devel project having the latest cobbler package building against Factory/latest as soon as this works, enable building against older distro XY. One can then link in missing package dependencies and make them also build against XY and cobbler can use them for its dependencies.
Not sure I explained this in an understandable way or whether this could help at all, as you probably know what you are doing on Centos (I do not)...

I agree with manpage creation part. Ideally would certainly be in setup.py and being transparent
to .spec as it was before (being build and installed via setup.py)

FYI, I found this bug inside the docs/Makefile. Variables are not overridable via commandline. I solved it on shortterm via ?=, override := could probably also be used, maybe this won't survive anyway...:

 # You can set these variables from the command line.
-SPHINXOPTS    =
-SPHINXBUILD   = sphinx-build
-PAPER         =
-BUILDDIR      = _build
+SPHINXOPTS    ?=
+SPHINXBUILD   ?= sphinx-build
+PAPER         ?=
+BUILDDIR      ?= _build

**Comment by rbberger** [comment]
@opoplawski ok, good to know. Thanks for clearing that up. In that case it makes sense to say CentOS 7 stays with the 2.x line and instead all energy should be directed to making the Python 3 port work as good as possible on Fedora 29+ / CentOS 8 / CentOS Stream / RHEL 8 and SUSE.

@watologo1 CentOS 8 is not an option yet for our production systems. It just came out and we're currently exploring the transition path for a possible migration from CentOS 7 to 8 sometime in 2020. This includes looking at the changes in cobbler since we use it for our provisioning. However, the current state of the cobbler master seems quite fragile. I so far ran into multiple outdated code paths during basic `cobbler import` and `cobbler sync` that seem to be due to the modules refactoring (like #2145).

FWIW, I've created a first version of a CentOS 8 compatible Spec file and Dockerfile based on the current ones in the repo. See https://github.com/rbberger/cobbler/commits/centos8_rpms. Besides a few packages that were renamed, the real blockers were outdated lines for the man pages and some directories/files that don't exist. If you want me to create a PR for this, I'll do so.

The difference I see between the spec file in the repo and the ones in OBS and COPR is that it works/worked on multiple distros. IMHO it's a benefit of having the spec file directly in the project git repo instead of having to hunt it down in different places. If maintaining it as one file becomes too much of a hassle, maybe having the latest of OBS and COPR spec files in a folder would be a compromise.

My next steps: I'll move forward with the CentOS 8 RPMs I created based on the current master and test a simple cluster setup on real hardware and document what I find.

Concerning the man pages build: Looks like Sphinx has a integration for [setuptools](https://www.sphinx-doc.org/en/master/usage/advanced/setuptools.html) that maybe could be used. Probably this should be its own issue.
